### PR TITLE
Support NNBD (Null safety) for generated code of mobx_codegen (new)

### PR DESCRIPTION
**Pull request by fzyzcjy**
Same code as https://github.com/mobxjs/mobx.dart/pull/631

### CODE REVIEW

File: mobx_codegen/lib/src/template/computed.dart [hunk 1]
```diff
@@ -10,7 +10,7 @@ class ComputedTemplate {
   @override
   // ignore: prefer_single_quotes
   String toString() => """
-  Computed<$type> $computedName;
+  Computed<$type>? $computedName;
```
Comment (pavanpodila) [inline mobx_codegen/lib/src/template/computed.dart:13]: Why are you forcing nullable here for the property ? The Computed is always non-nullable, instead the value it produces can potentially be nullable depending on how the user sets it up. Please remove the `?' here

File: mobx_codegen/lib/src/template/params.dart [hunk 2]
```diff
@@ -3,17 +3,16 @@ class ParamTemplate {
...
   String defaultValue;
   bool hasRequiredAnnotation = false;
+  bool hasRequiredKeyword = false;
 
   String get asArgument => name;
 
   NamedArgTemplate get asNamedArgument => NamedArgTemplate()..name = name;
 
-  String get metadata => hasRequiredAnnotation ? '@required ' : '';
+  String get metadata => (hasRequiredAnnotation ? '@required ' : '') + (hasRequiredKeyword ? 'required ' : '');
```
Comment (pavanpodila) [inline mobx_codegen/lib/src/template/params.dart:12]: I think the annotation is no longer valid in this version of Dart, let's just use the keyword check instead

**Review by pavanpodila — CHANGES_REQUESTED** [review]
much cleaner...just a few more changes

File: mobx_codegen/lib/src/template/computed.dart (hunk 1 above)
Comment (fzyzcjy) [reply mobx_codegen/lib/src/template/computed.dart:13]: But we need to support `$computedName ??= Computed()`, so it should be nullable

File: mobx_codegen/lib/src/template/params.dart (hunk 2 above)
Comment (fzyzcjy) [reply mobx_codegen/lib/src/template/params.dart:12]: ok

File: mobx_codegen/lib/src/template/computed.dart (hunk 1 above)
Comment (pavanpodila) [reply mobx_codegen/lib/src/template/computed.dart:13]: That is for the name of the computed property, not the Computed type

**Review by pavanpodila — CHANGES_REQUESTED** [review]

File: mobx_codegen/lib/src/template/computed.dart (hunk 1 above)
Comment (fzyzcjy) [reply mobx_codegen/lib/src/template/computed.dart:13]: I am confused... if remove "?" then error.

Comment (fzyzcjy) [reply mobx_codegen/lib/src/template/computed.dart:13]: Normally (notice the type is `int?`here, but often it is things like `int` in other places:

image


If no "?"

image

If use late:

image

**Comment by fzyzcjy** [comment]
@pavanpodila See the comments above

File: mobx_codegen/lib/src/template/computed.dart (hunk 1 above)
Comment (pavanpodila) [reply mobx_codegen/lib/src/template/computed.dart:13]: ya I see the issue now. Computed is created lazily on the first access, so I think late is the right approach. Not sure why greyed out text. Can you replace with `late` for now?

**Review by pavanpodila — CHANGES_REQUESTED** [review]

File: mobx_codegen/lib/src/template/computed.dart (hunk 1 above)
Comment (pavanpodila) [reply mobx_codegen/lib/src/template/computed.dart:13]: the initialization of the computed variable can be move to the declaration once we make it lazy. In fact we can make it `final lazy`

**Review by pavanpodila — CHANGES_REQUESTED** [review]

**Comment by fzyzcjy** [comment]
@pavanpodila Hmm when I do this, error:

image

**Comment by pavanpodila** [comment]
Aah looks like only this reference is allowed with late variables. Alright let's go back to nullable 👍

**Comment by fzyzcjy** [comment]
@pavanpodila So when can we release the new version ;)

**Review by pavanpodila — APPROVED** [review]

**Comment by pavanpodila** [comment]
> @pavanpodila So when can we release the new version ;)

I updated the sdk version to `2.12.0` and looks like I had a ton of errors on null-safety ;-). Will try to fix this today and hopefully a release in a day or so 🤔

**Comment by pavanpodila** [comment]
Sadly we cannot upgrade to 2.12.0, just yet because there are some dependencies that have not been migrated, esp. the build and source_gen packages. We will have to suppress the legacy import warning and make some changes to see if it works. 

Try changing the sdk version to `2.12.0` and see how far you can go ?

**Comment by fzyzcjy** [comment]
Let me have a look. Actually we may use 2.12 and solve the errors, while some dependencies are legacy?

**Comment by fzyzcjy** [comment]
See this: https://github.com/fzyzcjy/mobx.dart/tree/new_null_safety_2

Test can be run by `dart --no-sound-null-safety test `

**Comment by pavanpodila** [comment]
I'll give it a try later today or tomorrow. Have a bunch of client deliverables this week.

**Comment by hacker1024** [comment]
There are dependency conflicts with the latest version of `build_runner`.

```
Because every version of mobx_codegen from git depends on build_resolvers ^1.5.4 and build_runner >=1.12.0 depends on build_resolvers ^2.0.0, mobx_codegen from git is incompatible with build_runner >=1.12.0.
So, because app depends on both build_runner ^1.12.1 and mobx_codegen from git, version solving failed.
pub get failed (1; So, because app depends on both build_runner ^1.12.1 and mobx_codegen from git, version solving failed.)
```

**Pull request merged**